synapse/storage/databases/main/deviceinbox.py

```diff
@@ -664,21 +664,28 @@ def _remove_deleted_devices_from_device_inbox_txn(
             due to a single device having lots of device messages.
             """
 
-            last_stream_id = progress.get("stream_id", 0)
+            txn.execute("SELECT max(stream_id) FROM device_inbox")
```

Comment:
it might be nice to do this once when the update starts, and stash the result in the progress dict. That means we won't re-scan messages which arrive after the upgrade begins.